superset/security/manager.py

```diff
@@ -2063,6 +2064,27 @@ def can_access_based_on_dashboard(datasource: "BaseDatasource") -> bool:
         )
 
         exists = db.session.query(query.exists()).scalar()
+
+        # check for datasets that are only used by filters
```

Comment:
I think this logic makes sense. Note in https://github.com/apache/superset/pull/24789 this method is slated for removal, but the addition of the integration test will ensure that the logic will be preserved.

Reply:
@john-bodley I totally agree that this approach is far from ideal from a performance point of view. I think with the bigger changes that are going to be discussed, we can reformulate this process and make sure this validation happens with a target dashboard ID in mind. I saw in https://github.com/apache/superset/pull/24804 that you have updated the `raise_for_access` function so it can also receive a `dashboard`, so I think it would be easier to implement this improvement once those changes (and other decisions made in regards to expected behavior) are made.

Personally this is my second contribution to Superset so I would rather avoid doing bigger estructural changes until I get more familiar with the code as a whole.